### DISCUSSION TITLE
documentation on what javascript features are supported in design document view scripts?

### DISCUSSION THREAD

**Discussion by fangq** [apache/couchdb#4971] [General] [answered]
I am trying to write a few map functions to create views in a CouchDB database. Futon kept telling me that the script I wrote is not a valid function - despise that I used mostly standard JS features. After some debugging, I found that if I use standard `string.match(//)`, `string.replace(//,'')`, `array.filter()`, `array.map()` calls, or lambda functions `((x) => x * 2)`, it failed to save the document and threw an error.

This gives me the feeling that couchdb only supports a subset of JS, but could not find a document on what JS features are not supported.

Can someone let me know if such document exist? why couchjs does not support standard JS functions? from #1875 and #2345, it seems newer spidermonkey versions have been supported, but feel a bit surprised that it does not support these basic JS features.

my server has couchdb v3.3.2.

**Comment by rnewson** [comment]
This is determined by which version of spidermonkey your couchdb is built with.

**Reply by big-r81** [reply, marked as answer]
And you can show the used Spidermonkey engine with this [endpoint](https://docs.couchdb.org/en/stable/api/server/common.html#node-node-name-versions).

**Comment by fangq** [comment]
thank you both for your prompt replies! the output of `_node/_local/_versions` shows that my OS (Ubuntu 18.04) has an old version of spidermonkey

```json
"javascript_engine":{"version":"1.8.5","name":"spidermonkey"}
```

I will try couchdb on a newer server.
